### PR TITLE
Update product review graphql schema

### PR DESCRIPTION
**Pull request by sky-hub**
## Problem

1. Can not sort reviews by other fields
1. Reviews query does not return totals reviews
1. Unable to access review detail page from customer account as the route requires review id.
1. Keep field naming consistency.
1. Missing rating percent from review ratings.
1. Usage of "String" field type instead of "Int" when it is not the case.

## Solution

1. Review sort order
	- useful if we want to create a listing/widget with top 5 product reviews
	- default will be "created_at" DESC
	
1. Add "total_count" field for "ProductReviews"
	- needed to display review totals when listing product reviews in customer account
	
1. Add field "review_id" to "ProductReview"
	- needed if we want to keep the review detail page from customer account. See route "review/customer/view/id/[REVIEW_ID]" (Ex: https:www.example.com/review/customer/view/id/1/). If this page can be eliminated we can drop this field.
	
1. Rename "ProductReview" field "summary" with "title" and field "text" with "review_text"
	- this is to keep them alligned with https://github.com/magento/magento2/pull/19266/files
1. Rename "CreateProductReviewInput" field "summary" with "title" and field "text" with "review_text"
	- this is to keep them alligned with https://github.com/magento/magento2/pull/19266/files

1. Rename "ProductReview" field "ratings_breakdown" with "ratings".
	- keep naming consitency with same field from "CreateProductReviewInput"
	- keep field naming as intuitive as possible
	
1. Add percent field to "ProductReviewRating"
	- help to avoid calculations in frontend area if the value for this field is needed for frontend logic.
	
1. Change "ProductReviewRating" field type from "string" to "int".
	- there is not need to have field type as string when Magento Core predefined values are 1 to 5 and does not allows doe users to add new values.
	
1. Change "value_id" field from "ProductReviewRatingValueMetadata" to int instead of string
	- there is not need to have field type as string when Magento Core predefined values are 1 to 5 and does not allows doe users to add new values.

Note: Instead of base64 encoded values for ratings we can use the rating "Default Value" which is unique across the system, but that will mean adding a separate field for store label.

```graphql
type ProductReviewRatingMetadata {
    id: String! @doc(description: "Base 64 encoded rating id.")
    name: String! @doc(description: "The review rating name. For example Quality, Price")
    label: String! @doc(description: "The review rating store label.")
    values: [ProductReviewRatingValueMetadata!]! @doc(description: "List of product review ratings sorted based on position.")
}
```

instead of 
```graphql
type ProductReviewRatingMetadata {
    id: String! @doc(description: "Base 64 encoded rating id.")
    name: String! @doc(description: "The review rating name. For example Quality, Price")
    values: [ProductReviewRatingValueMetadata!]! @doc(description: "List of product review ratings sorted based on position.")
}
```

That would mean "ProductReviewRatingInput " will look like this:
```graphql
type ProductReviewRatingInput {
    name: String! @doc(description: "The review rating name. For example Quality, Price")
    value: Int! @doc(description: "The rating value. Possible values: 1, 2, 3, 4, 5")
}
```
	
## Requested Reviewers

@naydav 
@paliarush 
@lenaorobei

### CODE REVIEW

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 1]
```diff
@@ -45,13 +48,13 @@ type ProductReviewRatingsMetadata {
...
     id: String! @doc(description: "Base 64 encoded rating id.")
-    name: String! @doc(description: "The review rating name for example quality, price")
+    name: String! @doc(description: "The review rating name. For example Quality, Price")
     values: [ProductReviewRatingValueMetadata!]! @doc(description: "List of product review ratings sorted based on position.")
 }
 
 type ProductReviewRatingValueMetadata {
     value_id: String! @doc(description: "Base 64 encoded rating value id.")
-    value: String! @doc(description: "e.g Good, Perfect, 3, 4, 5")
+    value: Int!@doc(description: "The rating value. Possible values: 1, 2, 3, 4, 5")
```
Comment (akaplya) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]: product review rating does support custom values including strings

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 2]
```diff
@@ -65,12 +68,17 @@ type CreateProductReviewOutput {
 input CreateProductReviewInput {
     sku: String! @doc(description: "The SKU of the product that the review is assigned")
     nickname: String! @doc(description: "The customer's nickname. Defaults to customer name if logged in.")
-    summary: String! @doc(description: "The review summary (a.k.a title")
-    text: String! @doc(description: "The review text.")
-    ratings: [ProductReviewRatingInput!]! @doc(description: "Ratings details by category. e.g price: 5, quality: 4 etc")
+    title: String! @doc(description: "The review title")
+    review_text: String! @doc(description: "The review text.")
```
Comment (akaplya) [inline design-documents/graph-ql/coverage/product-reviews.graphqls]: what the necessity of this change? smaller name informative enough I would avoid using snake_case fields if possible because of the nature of main graphql clients - js apps.

Comment (sky-hub) [reply design-documents/graph-ql/coverage/product-reviews.graphqls]: 1. Please see point 5 from PR description.
2. If the goal is to avoid snake_case fields please see point 6 also.

File: design-documents/graph-ql/coverage/product-reviews.graphqls (hunk 1 above)
Comment (sky-hub) [reply design-documents/graph-ql/coverage/product-reviews.graphqls]: I did not know there is such support.

**Comment by sky-hub** [comment]
@lenaorobei , can I get an update on these changes? Are they still of interest? Are there any additional changes/updates required?

Thanks

File: design-documents/graph-ql/coverage/product-reviews.graphqls (hunk 1 above)
Comment (akaplya) [reply design-documents/graph-ql/coverage/product-reviews.graphqls:57]: It is technically possible through the rating configuration so let's preserve this behavior in the schema.

File: design-documents/graph-ql/coverage/product-reviews.graphqls (hunk 2 above)
Comment (paliarush) [reply design-documents/graph-ql/coverage/product-reviews.graphqls:72]: We are designing these API from scratch, there is no need to be consistent with existing service contracts. 
We definitely need to make changes that make API more self-explanatory. 

I believe current naming of the field `ProductReview.text` is intuitive and avoids repetitions compared to `ProductReview.review_text`.
`ProductReview.title` vs `ProductReview.summary` is pretty subjective. When we were selecting name for this field we chose `summary` because it it not an actual review title, but rather a short summary.

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 3]
```diff
@@ -11,30 +11,33 @@ type ProductInterface {
...
 }
 
 type ProductReviews {
     items: [ProductReview]! @doc(description: "An array of product reviews.")
     page_info: SearchResultPageInfo! @doc(description: "Metadata for pagination rendering.")
+    total_count: Int @doc(description: "The number of product reviews returned")
 }
 
 type ProductReview @doc(description: "Details of a product review") {
+    review_id: Int! @doc(description: "The ID assigned to the review")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls:25]: Based on latest agreements should be:
```suggestion
    uid: ID! @doc(description: "The ID assigned to the review")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 4]
```diff
@@ -11,30 +11,33 @@ type ProductInterface {
...
     product: ProductInterface! @doc(description: "Contains details about the reviewed product")
-    summary: String! @doc(description: "The review summary (a.k.a title")
-    text: String! @doc(description: "The review text.")
+    title: String! @doc(description: "The review title")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls:27]: Based on https://github.com/magento/architecture/pull/343#discussion_r461735711 
let's revert this change.

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 5]
```diff
@@ -11,30 +11,33 @@ type ProductInterface {
...
     average_rating: Float! @doc(description: "The average rating for product review.")
-    ratings_breakdown: [ProductReviewRating!]! @doc(description: "An array of ratings by rating category. For example quality, price.")
+    ratings: [ProductReviewRating!]! @doc(description: "An array of ratings by rating category. For example Quality, Price.")
 }
 
 type ProductReviewRating {
-    name: String! @doc(description: "The review rating name for example quality, price.")
-    value: String! @doc(description: "The rating value given by customer. Possible values by default: 1 to 5.")
+    name: String! @doc(description: "The review rating name. For example Quality, Price.")
+    percent: Float! @doc(description: "The rating percent given by customer to the review")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls:37]: Based on https://github.com/magento/architecture/pull/343#discussion_r356165700 let's make it optional.
```suggestion
    percent: Float @doc(description: "The rating percent given by customer to the review")
```

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 6]
```diff
@@ -11,30 +11,33 @@ type ProductInterface {
...
+    value: Int! @doc(description: "The rating value given by customer. Possible values by default: 1 to 5.")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls:38]: Based on https://github.com/magento/architecture/pull/343#discussion_r356165700 we should revert this change.

File: design-documents/graph-ql/coverage/product-reviews.graphqls (hunk 1 above)
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls:57]: Based on https://github.com/magento/architecture/pull/343#discussion_r356165700 it should be reverted

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 7]
```diff
@@ -65,12 +68,17 @@ type CreateProductReviewOutput {
 input CreateProductReviewInput {
     sku: String! @doc(description: "The SKU of the product that the review is assigned")
     nickname: String! @doc(description: "The customer's nickname. Defaults to customer name if logged in.")
-    summary: String! @doc(description: "The review summary (a.k.a title")
-    text: String! @doc(description: "The review text.")
-    ratings: [ProductReviewRatingInput!]! @doc(description: "Ratings details by category. e.g price: 5, quality: 4 etc")
+    title: String! @doc(description: "The review title")
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls:71]: Based on https://github.com/magento/architecture/pull/343#discussion_r461735711 changes in these two lines should be reverted.

File: design-documents/graph-ql/coverage/product-reviews.graphqls [hunk 8]
```diff
@@ -65,12 +68,17 @@ type CreateProductReviewOutput {
...
+    review_text: String! @doc(description: "The review text.")
+    ratings: [ProductReviewRatingInput!]! @doc(description: "Ratings details by category. e.g Price: 5, Quality: 4, etc")
 }
 
 type ProductReviewRatingInput {
     id: String! @doc(description: "Base 64 encoded rating id.")
     value_id: String! @doc(description: "Base 64 encoded rating value id.")
 }
+
+input ReviewSortInput @doc(description: "Specifies the field to use for sorting results and indicates whether the results are sorted in ascending or descending order") {
```
Comment (paliarush) [inline design-documents/graph-ql/coverage/product-reviews.graphqls:81]: @nrkapoor please confirm we need this feature

**Comment by nrkapoor** [comment]
@sky-hub @paliarush I have added it to the backlog and will prioritize with my team. Closing this PR and will open another one once we determine the priority.

**Comment by sky-hub** [comment]
@nrkapoor , @paliarush , ok, just let me know if there is anything else i can help with on this matter.

Cheers
